File: project/.sdk/tm/rb/utility/fetcher.rb
# ProjectName SDK utility: fetcher
require_relative 'struct/voxgig_struct'
require 'net/http'
require 'uri'
require 'json'

module ProjectNameUtilities
  DefaultHttpFetch = ->(fullurl, fetchdef) {
    method_str = fetchdef["method"] || "GET"
    body_str = fetchdef["body"]
    headers = fetchdef["headers"] || {}

    begin
      uri = URI.parse(fullurl)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == "https")

      klass = case method_str.upcase
              when "POST" then Net::HTTP::Post
              when "PUT" then Net::HTTP::Put
              when "DELETE" then Net::HTTP::Delete
              when "PATCH" then Net::HTTP::Patch
              else Net::HTTP::Get
              end

      request = klass.new(uri)
      has_ua = false
      headers.each do |k, v|
        next unless v.is_a?(String)
        has_ua = true if k.to_s.downcase == 'user-agent'
        request[k] = v.to_s
      end
      # Default User-Agent — Net::HTTP sets "Ruby" which some CDNs block.
      # Use a Mozilla-shaped UA unless the caller already set one.
      request['User-Agent'] = 'Mozilla/5.0 (compatible; ProjectNameSDK/1.0)' unless has_ua
      request.body = body_str if body_str.is_a?(String)

      resp = http.request(request)
      resp_headers = {}
      resp.each_header { |k, v| resp_headers[k.downcase] = v }

      json_body = nil
      begin
        json_body = JSON.parse(resp.body) if resp.body && !resp.body.empty?
      rescue JSON::ParserError
      end

      return {
        "status" => resp.code.to_i,
        "statusText" => resp.message,
        "headers" => resp_headers,
        "json" => -> { json_body },
        "body" => resp.body,
      }, nil
    rescue StandardError => e
      # Network-level failures (DNS, TCP, TLS, timeouts) — return a synthesized
      # response with status 0 so callers can branch on result.ok like any
      # other failed request, instead of seeing an unhandled exception.
      return {
        "status" => 0,
        "statusText" => "#{e.class}: #{e.message}",
        "headers" => {},
        "json" => -> { nil },
        "body" => nil,
      }, nil
    end
  }

  Fetcher = ->(ctx, fullurl, fetchdef) {
    if ctx.client.mode != "live"
      return nil, ctx.make_error("fetch_mode_block",
        "Request blocked by mode: \"#{ctx.client.mode}\" (URL was: \"#{fullurl}\")")
    end

    options = ctx.client.options_map
    if VoxgigStruct.getpath(options, "feature.test.active") == true
      return nil, ctx.make_error("fetch_test_block",
        "Request blocked as test feature is active (URL was: \"#{fullurl}\")")
    end

    sys_fetch = VoxgigStruct.getpath(options, "system.fetch")

    return DefaultHttpFetch.call(fullurl, fetchdef) if sys_fetch.nil?
    return sys_fetch.call(fullurl, fetchdef) if sys_fetch.is_a?(Proc)

    return nil, ctx.make_error("fetch_invalid", "system.fetch is not a valid function")
  }
end

File: project/.sdk/tm/php/utility/Fetcher.php
<?php
declare(strict_types=1);

// ProjectName SDK utility: fetcher

class ProjectNameFetcher
{
    public static function defaultHttpFetch(string $fullurl, array $fetchdef): array
    {
        $method_str = $fetchdef['method'] ?? 'GET';
        $body_str = $fetchdef['body'] ?? null;
        $headers = $fetchdef['headers'] ?? [];

        $opts = [
            'http' => [
                'method' => strtoupper($method_str),
                'ignore_errors' => true,
            ],
        ];

        $header_lines = [];
        foreach ($headers as $k => $v) {
            if (is_string($v)) {
                $header_lines[] = "{$k}: {$v}";
            }
        }

        if (is_string($body_str)) {
            $opts['http']['content'] = $body_str;
        }

        if (!empty($header_lines)) {
            $opts['http']['header'] = implode("\r\n", $header_lines);
        }

        $context = stream_context_create($opts);
        $response_body = @file_get_contents($fullurl, false, $context);

        $status = 0;
        $status_text = '';
        $resp_headers = [];

        if (isset($http_response_header) && is_array($http_response_header)) {
            foreach ($http_response_header as $header) {
                if (preg_match('/^HTTP\/\S+\s+(\d+)\s+(.*)$/i', $header, $matches)) {
                    $status = (int)$matches[1];
                    $status_text = trim($matches[2]);
                } else {
                    $parts = explode(':', $header, 2);
                    if (count($parts) === 2) {
                        $resp_headers[strtolower(trim($parts[0]))] = trim($parts[1]);
                    }
                }
            }
        }

        $json_body = null;
        if ($response_body !== false && $response_body !== '') {
            $decoded = json_decode($response_body, true);
            if (json_last_error() === JSON_ERROR_NONE) {
                $json_body = $decoded;
            }
        }

        return [
            [
                'status' => $status,
                'statusText' => $status_text,
                'headers' => $resp_headers,
                'json' => function () use ($json_body) { return $json_body; },
                'body' => $response_body !== false ? $response_body : '',
            ],
            null,
        ];
    }

    public static function call(ProjectNameContext $ctx, string $fullurl, array $fetchdef): array
    {
        if ($ctx->client->mode !== 'live') {
            return [null, $ctx->make_error('fetch_mode_block',
                "Request blocked by mode: \"{$ctx->client->mode}\" (URL was: \"{$fullurl}\")")];
        }

        $options = $ctx->client->options_map();
        if (\Voxgig\Struct\Struct::getpath($options, 'feature.test.active') === true) {
            return [null, $ctx->make_error('fetch_test_block',
                "Request blocked as test feature is active (URL was: \"{$fullurl}\")")];
        }

        $sys_fetch = \Voxgig\Struct\Struct::getpath($options, 'system.fetch');

        if ($sys_fetch === null) {
            return self::defaultHttpFetch($fullurl, $fetchdef);
        }
        if (is_callable($sys_fetch)) {
            return $sys_fetch($fullurl, $fetchdef);
        }

        return [null, $ctx->make_error('fetch_invalid', 'system.fetch is not a valid function')];
    }
}
